Update the Asset Pipeline guide

matuszewskijan opened the pull request:
<!--
Thanks for contributing to Rails!

Please do not make *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.

Create a pull request when it is ready for review and feedback
from the Rails team :).

If your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

### Motivation / Background

<!--
Describe why this Pull Request needs to be merged. What bug have you fixed? What feature have you added? Why is it important?
If you are fixing a specific issue, include "Fixes #ISSUE" (replace with the issue number, remove the quotes) and the issue will be linked to this PR.
-->

This Pull Request has been created because the actual asset pipeline guide were outdated in many ways:
- It was likely written for Rails 4, a lot of stuff not relevant anymore
- Not mentioning the `importmaps-rails` as the default solution for JavaScript
- The sprockets guide was written before `Sprockets 4.0` so it never mentioned the `manifest.js` file and some parts were not relevant anymore
- The `jsbundling-rails`, `cssbundling-rails`, `dartsass-rails` and `tailwindcss-rails` gems weren't mentioned at all

### Detail

This Pull Request changes the [Asset Pipeline](https://guides.rubyonrails.org/asset_pipeline.html) guide to reflect the Rails 7 asset pipeline configuration.

### Additional information

I've started my work by taking the https://github.com/rails/rails/pull/45400 draft PR. But in fact it wasn't too helpful and in the end, I've rewritten almost everything by myself.

For the Importmap section, I've mostly copied the Readme from the `importmap-rails` gem.

Sprockets section changes:
- tried to avoid showing that it could also handle the JavaScript to not confuse the readers, I am not sure if the new version is clear enough?
- There was a lot of info about older Sprockets versions and I did my best to find and update all of the outdated parts, is the refreshed version still correct and understandable?
- I've also removed all of the SCSS mentions there as we should point anyone interested in Sass to use `dartsass-rails` instead of the unmaintained `sassc-rails` gem
- Also removed all of the `CoffeeScript` mentions

About the last `Alternative Libraries` section: I've tried to only give a clue for the readers to show that there are other resources that might be useful for them. Isn't it too short?

First time contributing so I'd highly appreciate any feedback!


### Checklist

Before submitting the PR make sure the following are checked:

* [X] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [X] Tests are added or updated if you fix a bug or add a feature.
* [X] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

matuszewskijan committed:
Rewrite Asset Pipeline guide for Rails 7

Add back Sprockets guide(with remarks) to Asset Pipeline guide

Describe Import Maps in asset pipeline guide

Small rewording to first chapters of Asset Pipeline guide

Remove Sass and CoffeScript from Codin Links to Assets section

Remove Preprocessing section from Asset Pipeline guide

Mention dartsass-rails and rewrite beginning Asset Pipeline guide

Add more details on `rails-importmaps` in Asset Pipeline guide

Add alternative libraries section

Align Asset Pipeline guide Sprockets section for Sprockets V4

Improve the introduction to Asset Pipeline guide

Fix formatting in Importmap-rails section of Asset Pipeline guide

Reword, fix typos and add more context to the Asset Pipeline guide

Style, formatting and cleanup changes for asset pipeline guide

Minor style/typo fixes to Asset Pipeline guide

Remove sections dobuled by mistake from asset pipeline guide

Co-authored-by: Alex Ghiculescu <alex@tanda.co>
Co-authored-by: Breno Gazzola <breno.gazzola@gmail.com>
Co-authored-by: Shozo Hatta <hachi8833@gmail.com>

ghiculescu commented on guides/source/asset_pipeline.md:
```diff
@@ -3,87 +3,63 @@
-other languages and pre-processors such as CoffeeScript, Sass, and ERB.
-It allows assets in your application to be automatically combined with assets
-from other gems.
+The asset pipeline provides a framework to handle the delivery of Javascript and
```
```suggestion
The asset pipeline provides a framework to handle the delivery of JavaScript and
```

ghiculescu commented:
This is awesome. When you are happy with it, could you squash down to a single commit please.

ghiculescu commented on guides/source/asset_pipeline.md:
```diff
@@ -100,194 +76,289 @@ content into the name, usually at the end. For example a CSS file `global.css`
 
-* [Optimize caching](https://developers.google.com/speed/docs/insights/LeverageBrowserCaching)
-* [Revving Filenames: don't use querystring](http://www.stevesouders.com/blog/2008/08/23/revving-filenames-dont-use-querystring/)
+This is **not compatible** with the ES Module loader spec. Instead, **to be ESM compatible**, you must provide 1 of the 3 following types of specifiers:
```
I think for a first time user this is probably a bit technical, it's talking about concepts they may have never come across or don't care about.

For someone coming from a package.json world it's great, but I don't know if it needs to be the first thing you see here.

matuszewskijan replied:
I've updated this part to show import maps more like an "alias" for the library source address, as it's the simplest definition of an import map I could think of. What do you think about the newer version?

ghiculescu commented on guides/source/asset_pipeline.md:
```diff
@@ -100,194 +76,289 @@ content into the name, usually at the end. For example a CSS file `global.css`
 
-Besides the standard `assets/*` paths, additional (fully qualified) paths can be
-added to the pipeline in `config/initializers/assets.rb`. For example:
+To avoid the waterfall effect where the browser has to load one file after another before it can get to the deepest nested import, importmap-rails supports [modulepreload links](https://developers.google.com/web/updates/2017/12/modulepreload). Pinned modules can be preloaded by appending `preload: true` to the pin.
```
Is there some generic advice we can give on which modules you should/shouldn't preload?

matuszewskijan replied:
I did some Googling and I didn't find anything super useful, most clarifying for me was this answer:
https://stackoverflow.com/a/63924789/9984998

And I am thinking of adding such a sentence to the guide:
> Preloading is extremely useful in case you know that you'd need a specific part of the code, but you don't want to execute it immediately.

Getting to specific about the library might be difficult, what do you think?

ghiculescu replied:
How about this:

> It's a good idea to preload libraries or frameworks that are used throughout your app, as this will tell the browser to download them sooner.

Stimulus and Turbo get preloaded by default

https://github.com/hotwired/stimulus-rails/blob/498902a3687d32bd6589f7d3c7e7d72eabd822c0/lib/install/stimulus_with_importmap.rb#L15

And I think this explanation is okay for why.

ghiculescu commented on guides/source/asset_pipeline.md:
```diff
@@ -100,194 +76,289 @@ content into the name, usually at the end. For example a CSS file `global.css`
+How to Use Sprockets
+-----------------------------
+
+In older versions of Rails, all assets were located in subdirectories of
```
I personally find it unhelpful when you read the current Rails guides and they say "older versions of Rails did X". Particularly when working on a new app, I don't care - I want to know the best way to do things now and into the future.

My preference is to have a note about upgrading/compatibility and have it marked as such for people who need to know.

matuszewskijan replied:
Rewritten trying to show why Sprockets is helpful and what kind of problems it solve:
> The naive approach to expose your application assets to the web would be to store them in
subdirectories of `public` folder such as `images` and `stylesheets`. Doing so manually would be difficult as most of the modern web applications require the assets to be processed in specific way for eg. compressing and adding fingerprints to the assets.

Sprockets is designed to automatically preprocess your assets stored in the configured directories and after processing expose them in the `public/assets` folder with fingerprinting, compression, source maps generation and other configurable features.

ghiculescu commented on guides/source/asset_pipeline.md:
```diff
@@ -100,194 +76,289 @@ content into the name, usually at the end. For example a CSS file `global.css`
+
+This is meant to include the contents of all files found in the `./app/assets/images` directory or any subdirectories as well as any file recognized as JS directly at `./app/javascript` or `./vendor/javascript`.
+
+It will load any CSS from the `./app/assets/stylesheets` directory(not including subdirectories). Assuming that you have `application.css` and `marketing.css` files in the `./app/assets/stylesheets` folder it will allow you to load those stylesheets with `<%= stylesheet_link_tag "application" %>` or `<%= stylesheet_link_tag "marketing" %>` from your views.
```
```suggestion
It will load any CSS from the `./app/assets/stylesheets` directory (not including subdirectories). Assuming that you have `application.css` and `marketing.css` files in the `./app/assets/stylesheets` folder it will allow you to load those stylesheets with `<%= stylesheet_link_tag "application" %>` or `<%= stylesheet_link_tag "marketing" %>` from your views.
```

ghiculescu commented on guides/source/asset_pipeline.md:
```diff
@@ -1195,3 +1041,51 @@ it as a preprocessor for your MIME type.
+
+Webpacker has been the default JavaScript pre-processor and bundler for Rails 5 and 6. `Webpacker` successor is now [`shakapacker`](https://github.com/shakacode/shakapacker) and it's advised to use it.
+
+The `webpacker`/`shakapacker` unlike other libraries in this list is completely independent of Sprockets and could process both JavaScript and CSS files. We encourage you to read the [Webpacker guide](https://guides.rubyonrails.org/webpacker.html) to learn more.
```
```suggestion
Unlike other libraries in this list, `webpacker`/`shakapacker` is completely independent of Sprockets and could process both JavaScript and CSS files. Read the [Webpacker guide](https://guides.rubyonrails.org/webpacker.html) to learn more.
```

ghiculescu commented on guides/source/asset_pipeline.md:
```diff
@@ -1195,3 +1041,51 @@ it as a preprocessor for your MIME type.
+
+### [Webpacker/Shakapacker](https://guides.rubyonrails.org/webpacker.html)
+
+Webpacker has been the default JavaScript pre-processor and bundler for Rails 5 and 6. `Webpacker` successor is now [`shakapacker`](https://github.com/shakacode/shakapacker) and it's advised to use it.
```
```suggestion
Webpacker was the default JavaScript pre-processor and bundler for Rails 5 and 6. It has now been retired. A successor called [`shakapacker`](https://github.com/shakacode/shakapacker) exists, but is not maintained by the Rails team or project.
```

matuszewskijan replied:
The first line of README in Shakapacker repo goes like this:
> Official, actively maintained successor to [rails/webpacker](https://github.com/rails/webpacker).

If it's not maintained by Rails team then maybe they shouldn't use the `official` word, but anyway I like the rewording there and going to commit it.

ghiculescu replied:
I'm not sure what official means in that context. It was added in https://github.com/shakacode/shakapacker/pull/14

I suggested that wording because it's not part of the `rails/` github org so isn't owned by the same core team.

cc @justin808 you may have some thoughts on this (or on this PR in general).

justin808 replied:
It's official in the sense that @dhh recommends it as the successor, and he came up with the name "shakapacker."

See https://github.com/rails/webpacker.

ghiculescu commented on guides/source/asset_pipeline.md:
```diff
@@ -1195,3 +1041,51 @@ it as a preprocessor for your MIME type.
+
+**What's the difference between Sprockets?** Sprockets on their own is not able to transpile the Sass into CSS, Node.js is required to generate the `.css` files from your `.sass`  files. Once the `.css` files are generated then `Sprockets` is able to deliver them to your clients.
+
+NOTE: `cssbundling-rails` is relying on Node to process the CSS. The `dartsass-rails` and `tailwindcss-rails` gems are using the standalone versions of Tailwind CSS and Dart Sass meaning no Node dependency. If you are using `importmap-rails` to handle your Javascripts and `dartsass-rails` or `tailwindcss-rails` for CSS you could completely avoid the Node dependency resulting in a less complex sol ...
```
```suggestion
NOTE: `cssbundling-rails` relies on Node to process CSS. The `dartsass-rails` and `tailwindcss-rails` gems use standalone versions of Tailwind CSS and Dart Sass, meaning no Node dependency. If you are using `importmap-rails` to handle your Javascripts and `dartsass-rails` or `tailwindcss-rails` for CSS you could completely avoid the Node dependency resulting in a less complex solution.
```

ghiculescu commented on guides/source/asset_pipeline.md:
```diff
@@ -1195,3 +1041,51 @@ it as a preprocessor for your MIME type.
+
+### [dartsass-rails](https://github.com/rails/dartsass-rails)
+
+If you are willing to use [`Sass`](https://sass-lang.com/) in your application `dartsass-rails` comes as a replacement for the legacy `sassc-rails` gem. `dartsass-rails` is using the `Dart Sass` implementation in favour of deprecated in 2020 [`LibSass`](https://sass-lang.com/blog/libsass-is-deprecated) used by `sassc-rails`.
```
```suggestion
If you want to use [`Sass`](https://sass-lang.com/) in your application, `dartsass-rails` comes as a replacement for the legacy `sassc-rails` gem. `dartsass-rails` uses the `Dart Sass` implementation in favour of deprecated in 2020 [`LibSass`](https://sass-lang.com/blog/libsass-is-deprecated) used by `sassc-rails`.
```

ghiculescu commented:
I think the only other notable thing missing from this guide is https://github.com/rails/propshaft

matuszewskijan commented:
> I think the only other notable thing missing from this guide is https://github.com/rails/propshaft

I asked a few questions on Rails Discord before posting this PR, and one of them was about Propshaft. DHH answered that it's a Rails 8 thing so I believe we could skip it for now. https://discord.com/channels/849034466856665118/987190327616090182/1061971064449675264

> This is awesome. When you are happy with it, could you squash down to a single commit please.

I will squash it all, let's wait a few more days as I posted it on weekend so maybe there will be more feedback and more commits to come.

zzak commented:
This is huge. Thanks for working on it!

![giphy](https://user-images.githubusercontent.com/277819/212575468-caca5ca4-e3af-4c16-b0ee-c7be62a16f50.gif)

matuszewskijan commented:
> This is awesome. When you are happy with it, could you squash down to a single commit please.

Done.

hachi8833 commented:
Awesome! 
We'd be glad if you merge this PR to https://github.com/rails/rails/tree/7-0-stable branch as well because current asset_pipeline.md in 7-0-stable is pretty confusing while translating.

zzak commented:
An error occurred when trying to build this locally, can you take a look? I will try to debug it during my day tomorrow if not:

```
Generating asset_pipeline.md as asset_pipeline.html
/Users/zzak/code/rails/guides/rails_guides/markdown.rb:47:in `dom_id': undefined method `[]' for nil:NilClass (NoMethodError)

          dom_id = "#{nodes[-2][:id]}-#{dom_id}"
                               ^^^^^
        from /Users/zzak/code/rails/guides/rails_guides/markdown.rb:115:in `block (2 levels) in generate_structure'
        from /Users/zzak/.gem/ruby/3.3.0/gems/nokogiri-1.13.10/lib/nokogiri/xml/node_set.rb:234:in `block in each'
        from /Users/zzak/.gem/ruby/3.3.0/gems/nokogiri-1.13.10/lib/nokogiri/xml/node_set.rb:233:in `upto'
        from /Users/zzak/.gem/ruby/3.3.0/gems/nokogiri-1.13.10/lib/nokogiri/xml/node_set.rb:233:in `each'
        from /Users/zzak/code/rails/guides/rails_guides/markdown.rb:100:in `block in generate_structure'
        from <internal:kernel>:90:in `tap'
        from /Users/zzak/code/rails/guides/rails_guides/markdown.rb:97:in `generate_structure'
        from /Users/zzak/code/rails/guides/rails_guides/markdown.rb:29:in `render'
        from /Users/zzak/code/rails/guides/rails_guides/generator.rb:160:in `generate_guide'
        from /Users/zzak/code/rails/guides/rails_guides/generator.rb:85:in `block in generate_guides'
        from /Users/zzak/code/rails/guides/rails_guides/generator.rb:83:in `each'
        from /Users/zzak/code/rails/guides/rails_guides/generator.rb:83:in `generate_guides'
        from /Users/zzak/code/rails/guides/rails_guides/generator.rb:40:in `generate'
        from rails_guides.rb:30:in `<main>'
```

You should be able to reproduce it by running `bundle exec rake guides:generate` from the guides directory.

matuszewskijan commented:
> An error occurred when trying to build this locally, can you take a look? I will try to debug it during my day tomorrow if not:
> 
> ```
> Generating asset_pipeline.md as asset_pipeline.html
> /Users/zzak/code/rails/guides/rails_guides/markdown.rb:47:in `dom_id': undefined method `[]' for nil:NilClass (NoMethodError)
> 
>           dom_id = "#{nodes[-2][:id]}-#{dom_id}"
>                                ^^^^^
>         from /Users/zzak/code/rails/guides/rails_guides/markdown.rb:115:in `block (2 levels) in generate_structure'
>         from /Users/zzak/.gem/ruby/3.3.0/gems/nokogiri-1.13.10/lib/nokogiri/xml/node_set.rb:234:in `block in each'
>         from /Users/zzak/.gem/ruby/3.3.0/gems/nokogiri-1.13.10/lib/nokogiri/xml/node_set.rb:233:in `upto'
>         from /Users/zzak/.gem/ruby/3.3.0/gems/nokogiri-1.13.10/lib/nokogiri/xml/node_set.rb:233:in `each'
>         from /Users/zzak/code/rails/guides/rails_guides/markdown.rb:100:in `block in generate_structure'
>         from <internal:kernel>:90:in `tap'
>         from /Users/zzak/code/rails/guides/rails_guides/markdown.rb:97:in `generate_structure'
>         from /Users/zzak/code/rails/guides/rails_guides/markdown.rb:29:in `render'
>         from /Users/zzak/code/rails/guides/rails_guides/generator.rb:160:in `generate_guide'
>         from /Users/zzak/code/rails/guides/rails_guides/generator.rb:85:in `block in generate_guides'
>         from /Users/zzak/code/rails/guides/rails_guides/generator.rb:83:in `each'
>         from /Users/zzak/code/rails/guides/rails_guides/generator.rb:83:in `generate_guides'
>         from /Users/zzak/code/rails/guides/rails_guides/generator.rb:40:in `generate'
>         from rails_guides.rb:30:in `<main>'
> ```
> 
> You should be able to reproduce it by running `bundle exec rake guides:generate` from the guides directory.

Ah sorry there was one conflict I resolved in the wrong way, it should be fixed now.

hachi8833 commented on guides/source/asset_pipeline.md:
```diff
@@ -18,44 +18,43 @@ After reading this guide, you will know:
+The fourth feature of the asset pipeline is it allows coding assets via a
+higher-level language for CSS.
 
 The first feature of the pipeline is to concatenate assets, which can reduce the
```
Looks like the four features are duplicate.

hachi8833 commented on guides/source/asset_pipeline.md:
````diff
@@ -18,44 +18,43 @@ After reading this guide, you will know:
-```ruby
-gem 'sassc-rails'
-```
+NOTE: This guide is focusing on the default asset pipeline using only `Sprockets` for CSS and `Importmaps-rails` for JavaScript processing. The main limitation of those two is that there is no support for transpiling so you can't use things like `Babel`, `Typescript`, `Sass`, `React JSX format` or `TailwindCSS`. We encourage you to read the [Alternative Libraries section](#alternative-libraries)  ...
````
- `Importmaps-rails`-> `importmap-rails`

hachi8833 commented on guides/source/asset_pipeline.md:
```diff
@@ -100,194 +99,278 @@ content into the name, usually at the end. For example a CSS file `global.css`
+Import Maps are the default Javascript processor, the logic of generating import maps is handled by the [`importmap-rails`](https://github.com/rails/importmap-rails) gem.
 
-2. **The file name can change between nodes in multi-server environments.**
+WARNING: Importmaps are used only for Javascript files and can not be used for CSS delivery. Check the [Sprockets section](#how-to-use-sprockets) to learn about CSS.
```
Importmaps -> Import maps

hachi8833 commented on guides/source/asset_pipeline.md:
```diff
@@ -100,194 +99,278 @@ content into the name, usually at the end. For example a CSS file `global.css`
-    When static assets are deployed with each new release of code, the mtime
-(time of last modification) of _all_ these files changes, forcing all remote
-clients to fetch them again, even when the content of those assets has not changed.
+Importmaps are essentially a string substitution for what is referred to as "bare module specifiers". They allow you to standardize the names of JavaScript module imports.
```
Importmaps -> import maps

hachi8833 commented on guides/source/asset_pipeline.md:
```diff
@@ -100,194 +99,278 @@ content into the name, usually at the end. For example a CSS file `global.css`
-will be served as static files by the application or web server when
-[`config.public_file_server.enabled`][] is set to true. You should use `app/assets` for
-files that must undergo some pre-processing before they are served.
+All of the configured import maps should be attached in `<head>` tag of your application by adding  `<%= javascript_importmap_tags %>` . The `javascript_importmap_tags` renders a bunch of scripts in the `head` tag:
```
- `<head>` tag -> `<head>` element
- `head` tag -> `<head>` element

hachi8833 commented on guides/source/asset_pipeline.md:
```diff
@@ -100,194 +99,278 @@ content into the name, usually at the end. For example a CSS file `global.css`
-ready to use by your application immediately using the `require_tree` directive. See
-[Manifest Files and Directives](#manifest-files-and-directives) for more details
-on require_tree.
+- Entrypoint for loading Javascript from `app/javascript/application.js`:
```
Javascript -> JavaScript

hachi8833 commented on guides/source/asset_pipeline.md:
```diff
@@ -100,194 +99,278 @@ content into the name, usually at the end. For example a CSS file `global.css`
 
-When doing this, ensure you are not using the `require_tree` directive, as that
-will result in your assets being included more than once.
+You can use the `./bin/importmap` command that's added as part of the `importmap-rails` install to pin, unpin, or update npm packages in your import map. The binstub uses `JSPM.org`
```
Just for convenience: `JSPM.org` -> [`JSPM.org`](https://jspm.org/)

hachi8833 commented on guides/source/asset_pipeline.md:
````diff
@@ -100,194 +99,278 @@ content into the name, usually at the end. For example a CSS file `global.css`
 ```
 
-would be referenced in a manifest like this:
+You can also pick an alternative, supported CDN provider when pinning, like `unpkg` or `jsdelivr` (`jspm` is the default):
````
Just for convenience: 
`unpkg` -> [`unpkg`](https://unpkg.com/)
`jsdelivr` -> [`jsdelivr`](https://www.jsdelivr.com/)

hachi8833 commented on guides/source/asset_pipeline.md:
```diff
@@ -100,194 +99,278 @@ content into the name, usually at the end. For example a CSS file `global.css`
+You can also opt to include controller-specific stylesheets files
+only in their respective controllers using the following:
+
+`<%= stylesheet_link_tag params[:controller] %>`
```
Wrapping with three backquotes with `html+erb` is preferable:

```html+erb
<%= stylesheet_link_tag params[:controller] %>
```

hachi8833 commented on guides/source/asset_pipeline.md:
```diff
@@ -1195,3 +1053,51 @@ it as a preprocessor for your MIME type.
+
+**When to use instead of `importmap-rails`?** If your JavaScript code depends on transpiling so if you are using [Babel](https://babeljs.io/), [TypeScript](https://www.typescriptlang.org/) or React `JSX` format then `jsbundling-rails` is the correct way to go.
+
+### [Webpacker/Shakapacker](https://guides.rubyonrails.org/webpacker.html)
```
FYI: while the situations and prospects are still very very unclear, it seems that webpack is being moving to the new Turbopack: 
https://vercel.com/blog/the-turbopack-vision

matuszewskijan commented:
@hachi8833 thank you for the review, I've fixed all of your points, and I've also found a few small typos and fixed them too. I've squashed all of the commits and rebased them with `main`.

> Awesome! We'd be glad if you merge this PR to https://github.com/rails/rails/tree/7-0-stable branch as well because current asset_pipeline.md in 7-0-stable is pretty confusing while translating. @rafaelfranca

According to this one, this branch is having conflicts with `7-0-stable` as I've rebased with the main. Shall I rebase it to `7-0`? I've created another branch in my local repo diverged from `7-0-stable` with the guide update commit(this branch is here: https://github.com/matuszewskijan/rails/tree/asset-pipeline-guide-7-0). I could open an PR with the secondary branch but I don't want to ping everyone so could you advise me how to do it best?

hachi8833 commented:
Thank you very much for the fixing! I appreciate for your works❤️

@matuszewskijan I think rebasing the branch to `7-0-stable` branch is OK and seems safer. I see that merging it would cause conflicts.

hachi8833 commented on guides/source/asset_pipeline.md:
```diff
@@ -1198,3 +1036,51 @@ it as a preprocessor for your MIME type.
+
+Webpacker was the default JavaScript pre-processor and bundler for Rails 5 and 6. It has now been retired. A successor called [`shakapacker`](https://github.com/shakacode/shakapacker) exists, but is not maintained by the Rails team or project.
+
+Unlike other libraries in this list `webpacker`/`shakapacker` is completely independent of Sprockets and could process both JavaScript and CSS files. Read the [Webpacker guide](https://guides.rubyonrails.org/webpacker.html) to learn more.
```
https://guides.rubyonrails.org/webpacker.html -> webpacker.html

matuszewskijan commented:
> Thank you very much for the fixing! I appreciate for your works❤️
> 
> @matuszewskijan I think rebasing the branch to `7-0-stable` branch is OK and seems safer. I see that merging it would cause conflicts.

Done, I've rebased this branch with `7-0-stable` and changed base branch of this pull request to `7-0-stable`. Hope it's all good, appreciate your comments!

zzak commented:
@matuszewskijan @hachi8833 I think what we want to do is get this merged to `main` first and then backport the merge commit. If there are conflicts between main and stable we can fix them then. Does that make sense?

hachi8833 commented:
@zzak yes, I should've explained more...

matuszewskijan commented:
> @matuszewskijan @hachi8833 I think what we want to do is get this merged to `main` first and then backport the merge commit. If there are conflicts between main and stable we can fix them then. Does that make sense?



> @zzak yes, I should've explained more...

No worries, it was actually my misunderstanding, I've changed the base branch to be again `main`.

hachi8833 commented:
Thank you for the changing branch!

zzak commented:
Was able to get this to build locally with matuszewskijan/rails#1, for anyone else interested in reviewing, I think this screenshot of the intro and chapter guide is a nice start:

![Screenshot 2023-03-22 at 12 27 38](https://user-images.githubusercontent.com/277819/226794759-50e0bcd5-17e3-4fe4-bf56-fb170dcda62d.png)

It does feel like Sprockets could be it's own guide at this point though 🤔

matuszewskijan commented:
> Was able to get this to build locally with [matuszewskijan#1](https://github.com/matuszewskijan/rails/pull/1), for anyone else interested in reviewing, I think this screenshot of the intro and chapter guide is a nice start:
> 
> ![Screenshot 2023-03-22 at 12 27 38](https://user-images.githubusercontent.com/277819/226794759-50e0bcd5-17e3-4fe4-bf56-fb170dcda62d.png)
> 
> It does feel like Sprockets could be it's own guide at this point though 🤔

Hey Zak, thanks for checking the branch, this section you deleted was supposed to be deleted, but I was rebasing the whole stuff many times and it seems that it sneaked back in the process. Thanks for fixing it!

Definitely, Sprockets could have its own guide. But maybe some of the basics are still useful there, still, a lot of stuff in the default setup is dependent on Sprockets.

zzak commented:
> Definitely, Sprockets could have its own guide. But maybe some of the basics are still useful there, still, a lot of stuff in the default setup is dependent on Sprockets.

Yeah, this _is_ the Sprockets guide, you're 💯 right. I'm not sure what was going through my head at the time :joy:.

Are you able to update your branch so that the guide builds here? I'd love to get this reviewed and merged. :pray:

hachi8833 commented:
@rafaelfranca it seems the PR is ready to merge. I hope you'd check this 🙏, apart from creating a new Sprockets guide.
cc: @zzak

zzak commented:
This needs a rebase, I'm happy to pitch in and help if that would move things along faster. Just want to make sure!

zzak commented:
Since I don't have edit access to @matuszewskijan's fork, I opened #47899 and merged. Thanks everyone for pitching in here!

matuszewskijan commented:
> Since I don't have edit access to @matuszewskijan's fork, I opened #47899 and merged. Thanks everyone for pitching in here!

Thank you @zzak, I am so happy to see my first contribution merged! Thanks everyone!
